h2o-core/src/main/java/water/parser/ARFFParser.java

```diff
@@ -127,55 +127,63 @@ private static int readArffHeader(int offset, ArrayList<String> header, byte[] b
     return offset;
   }
 
-  private static void processArffHeader(int ncols, String[] headerlines, String[][] data, String[] labels, String[][] domains, byte[] ctypes) {
+  static void processArffHeader(int ncols, String[] headerlines, String[] labels, String[][] domains, byte[] ctypes) {
     for (int i=0; i<ncols; ++i) {
-      data[i] = headerlines[i].split("\\s+");
-      if (!data[i][0].equalsIgnoreCase("@ATTRIBUTE")) {
+      String[] line = headerlines[i].split("\\s+", 2);
+      if (!line[0].equalsIgnoreCase(TAG_ATTRIBUTE)) {
         throw new ParseDataset.H2OParseException("Expected line to start with @ATTRIBUTE.");
       } else {
-        if (data[i].length < 3 ) {
+        final String spec = (line.length == 2) ? line[1].replaceAll("\\s", " ") : ""; // normalize separators
+        int sepIdx = spec.lastIndexOf(' ');
+        if (sepIdx < 0) {
           throw new ParseDataset.H2OParseException("Expected @ATTRIBUTE to be followed by <attribute-name> <datatype>");
         }
-        labels[i] = data[i][1];
-        String type = data[i][2];
+        final String type = spec.substring(sepIdx + 1).trim();
         domains[i] = null;
+        ctypes[i] = Vec.T_BAD;
         if (type.equalsIgnoreCase("NUMERIC") || type.equalsIgnoreCase("REAL") || type.equalsIgnoreCase("INTEGER") || type.equalsIgnoreCase("INT")) {
           ctypes[i] = Vec.T_NUM;
-          continue;
         }
         else if (type.equalsIgnoreCase("DATE") || type.equalsIgnoreCase("TIME")) {
           ctypes[i] = Vec.T_TIME;
-          continue;
         }
         else if (type.equalsIgnoreCase("ENUM")) {
           ctypes[i] = Vec.T_CAT;
-          continue;
         }
         else if (type.equalsIgnoreCase("STRING")) {
           ctypes[i] = Vec.T_STR;
-          continue;
         }
         else if (type.equalsIgnoreCase("UUID")) { //extension of ARFF
           ctypes[i] = Vec.T_UUID;
-          continue;
         }
         else if (type.equalsIgnoreCase("RELATIONAL")) {
           throw new UnsupportedOperationException("Relational ARFF format is not supported.");
         }
-        else if (type.startsWith("{") && data[i][data[i].length-1].endsWith("}")) {
-          StringBuilder builder = new StringBuilder();
-          for(int j = 2; j < data[i].length; j++) {
-            builder.append(data[i][j]);
-          }
-          domains[i] = builder.toString().replaceAll("[{}]", "").split(",");
-          if (domains[i][0].length() > 0) {
-            // case of {A,B,C} (valid list of factors)
-            ctypes[i] = Vec.T_CAT;
-            continue;
-          }
+        else if (type.endsWith("}")) {
```

Comment:
Regex would work too, what I need to find is the occurrence of pattern "{[^{]*}$" in the type specification. I can rewrite it with regex. Do you think it is worth it (do you see an issue with the current code)?